common/lessonplan/generate.ts

```diff
@@ -0,0 +1,298 @@
+    const tokens = enc.encode(text);
+    if (tokens.length <= maxTokens) {
+        enc.free();
+        return text;
+    }
+    const truncatedTokens = tokens.slice(0, maxTokens);
+    const truncatedText = new TextDecoder().decode(enc.decode(truncatedTokens));
+    enc.free();
+    return truncatedText;
+}
+
+// Function to resize image based on aspect ratio
+async function resizeImage(buffer: Buffer): Promise<Buffer> {
+    const image = sharp(buffer);
+    const metadata = await image.metadata();
+    const { width, height } = metadata;
+
+    if (!width || !height) {
+        throw new Error('Unable to determine image dimensions');
+    }
+
+    logger.debug(`Original image dimensions: ${width}x${height}`);
+
+    const tileSize = 512;
+    const aspectRatio = width / height;
+
+    logger.debug(`Image aspect ratio: ${aspectRatio}`);
+
+    let tilesX, tilesY;
+    if (aspectRatio > 1.1) {
+        // Landscape image
+        tilesX = 2;
+        tilesY = 1;
+        logger.debug(`Landscape image. Tiles: ${tilesX}x${tilesY}`);
+    } else if (aspectRatio < 0.9) {
+        // Portrait image
+        tilesX = 1;
+        tilesY = 2;
+        logger.debug(`Portrait image. Tiles: ${tilesX}x${tilesY}`);
+    } else {
+        // Square image
+        tilesX = tilesY = 1;
+        logger.debug(`Square image. Tiles: ${tilesX}x${tilesY}`);
+    }
+
+    const newWidth = tilesX * tileSize;
+    const newHeight = tilesY * tileSize;
+
+    logger.debug(`New image dimensions: ${newWidth}x${newHeight}`);
+    logger.debug(`Total tiles used: ${tilesX * tilesY}`);
+
+    return image.resize(newWidth, newHeight, { fit: 'contain', background: { r: 255, g: 255, b: 255, alpha: 1 } }).toBuffer();
+}
+
+export async function generateLessonPlan({
+    fileUuids,
+    subject,
+    grade,
+    duration,
+    state,
+    prompt,
+    expectedOutputs,
+    schoolType,
+    language = 'German', // Default language is German
+}: GenerateLessonPlanInput): Promise<Partial<z.infer<typeof plan>> & { subject: course_subject_enum; grade: string; duration: number }> {
+    logger.info(`Generating lesson plan for subject: ${subject}, grade: ${grade}, duration: ${duration} minutes, school type: ${schoolType}`);
+
+    const emptyFiles: string[] = [];
+    let combinedContent = '';
+    const imageContents: { type: 'image_url'; image_url: { url: string } }[] = [];
+
+    if (fileUuids.length > 0) {
+        // Fetch file contents for each UUID
+        const allFiles = await Promise.all(
+            fileUuids.map(async (uuid) => {
+                try {
+                    const file: File = getFile(uuid);
+                    const blob = new Blob([file.buffer], { type: file.mimetype });
+
+                    if (file.mimetype.startsWith('image/')) {
+                        // Handle image files
+                        logger.debug(`Processing image file: ${file.originalname}`);
+                        const resizedBuffer = await resizeImage(file.buffer);
```

Comment:
(I'm not sure what happens when you throw an APNG into this ... but I doubt someone will upload an APNG ;))